File: app/routes/arm.py
import json
import os
from flask import Blueprint, request, jsonify

from app.config import config
from app.services import get_control_service
from src.arm_control.angle_config import load_arm_angles, save_arm_angles

arm_bp = Blueprint("arm", __name__, url_prefix="/api/arm")
_DEFAULT_FILE = os.path.join(os.path.dirname(__file__), "..", "..", "arm_angles_default.json")


@arm_bp.route("/angles", methods=["GET", "POST"])
def arm_angles():
    driver = config.arm_driver

    if request.method == "GET":
        return jsonify({
            "driver": driver,
            "angles": load_arm_angles(driver),
        })

    payload = request.get_json(silent=True)
    if not isinstance(payload, dict):
        return jsonify({"error": "json body is required"}), 400

    body_driver = payload.get("driver", driver)
    if body_driver != driver:
        return jsonify({"error": f"driver mismatch: expected {driver}, got {body_driver}"}), 400

    angles_payload = payload.get("angles", payload)

    try:
        angles = save_arm_angles(driver, angles_payload)
        get_control_service().update_arm_angles(driver, angles)
    except ValueError as exc:
        return jsonify({"error": str(exc)}), 400

    return jsonify({
        "status": "success",
        "driver": driver,
        "angles": angles,
    })


@arm_bp.route("/angles/default", methods=["GET", "POST"])
def arm_angles_default():
    driver = config.arm_driver
    if request.method == "GET":
        try:
            with open(_DEFAULT_FILE) as f:
                data = json.load(f)
        except (FileNotFoundError, json.JSONDecodeError):
            return jsonify({"error": "default config not found"}), 404
        return jsonify({"driver": driver, "angles": data})

    # POST: 保存默认配置
    payload = request.get_json(silent=True)
    if not isinstance(payload, dict):
        return jsonify({"error": "json body required"}), 400
    angles_payload = payload.get("angles", payload)
    try:
        with open(_DEFAULT_FILE, "w") as f:
            json.dump(angles_payload, f)
    except IOError as e:
        return jsonify({"error": str(e)}), 500
    return jsonify({"status": "success", "driver": driver, "angles": angles_payload})


@arm_bp.route("/angles/preview", methods=["POST"])
def arm_angles_preview():
    driver = config.arm_driver
    payload = request.get_json(silent=True)
    if not isinstance(payload, dict):
        return jsonify({"error": "json body is required"}), 400

    body_driver = payload.get("driver", driver)
    if body_driver != driver:
        return jsonify({"error": f"driver mismatch: expected {driver}, got {body_driver}"}), 400

    key = payload.get("key")
    value = payload.get("value")
    angles_payload = payload.get("angles")

    if not isinstance(key, str):
        return jsonify({"error": "key is required"}), 400
    if value is None:
        return jsonify({"error": "value is required"}), 400
    if not isinstance(angles_payload, dict):
        return jsonify({"error": "angles is required"}), 400

    try:
        angle_value = int(value)
        angles = save_arm_angles(driver, angles_payload)
        service = get_control_service()
        service.update_arm_angles(driver, angles)
        service.preview_arm_angle(driver, key, angle_value)
    except ValueError as exc:
        return jsonify({"error": str(exc)}), 400

    return jsonify({
        "status": "success",
        "driver": driver,
        "key": key,
        "value": angle_value,
        "angles": angles,
    })

File: cpp/capp/src/routes/arm.cpp
// 机械臂角度
//
// 入口：register_arm_routes()（由 src/routes.cpp 的 register_routes 调用）
//
// 由 capp/src/routes.cpp 按域拆出来（对照 app/routes/*.py 的分法）。
// 对应 app/routes/arm.py。

#include "routes_internal.hpp"

#include <fstream>
#include <sstream>

namespace capp {
namespace routes {

// ── 机械臂角度 ──

void register_arm_routes(Router& router, AppContext& ctx) {


    // ── /api/arm ──
    router.add("GET", "/api/arm/angles", [&ctx](const HttpRequest&, HttpResponse& resp, ClientConn&, AppContext&) {
        Json j;
        j["driver"] = ctx.config.arm.backend;
        j["angles"] = csrc::load_arm_angles(ctx.config.arm.backend);
        resp.set_json(j);
    });

    router.add("POST", "/api/arm/angles", [&ctx](const HttpRequest& req, HttpResponse& resp, ClientConn&, AppContext&) {
        Json payload = req.json();
        if (!payload.is_object()) {
            resp.set_error("json body is required", 400);
            return;
        }
        std::string driver = ctx.config.arm.backend;
        if (payload.has("driver")) driver = payload.gets("driver");
        if (driver != ctx.config.arm.backend) {
            resp.set_error("driver mismatch: expected " + ctx.config.arm.backend + ", got " + driver, 400);
            return;
        }
        const Json* angles = payload.get("angles");
        if (!angles || !angles->is_object()) angles = &payload;

        Json normalized = csrc::save_arm_angles(driver, *angles);
        Json upd = update_arm_angles(ctx, driver, normalized);
        if (upd.has("error")) {
            resp.set_error(upd.gets("error"), 400);
            return;
        }
        Json j;
        j["status"] = "success";
        j["driver"] = driver;
        j["angles"] = normalized;
        resp.set_json(j);
    });

    // /api/arm/angles/default
    router.add("GET", "/api/arm/angles/default", [&ctx](const HttpRequest&, HttpResponse& resp, ClientConn&, AppContext&) {
        std::string path = csrc::arm_angles_path();
        size_t slash = path.find_last_of('/');
        std::string def_file = (slash == std::string::npos ? "" : path.substr(0, slash + 1)) + "arm_angles_default.json";
        std::ifstream f(def_file);
        if (!f) {
            resp.set_error("default config not found", 404);
            return;
        }
        std::stringstream ss;
        ss << f.rdbuf();
        Json data;
        if (!Json::parse(ss.str(), data)) {
            resp.set_error("default config parse failed", 404);
            return;
        }
        Json j;
        j["driver"] = ctx.config.arm.backend;
        j["angles"] = data;
        resp.set_json(j);
    });

    router.add("POST", "/api/arm/angles/default", [&ctx](const HttpRequest& req, HttpResponse& resp, ClientConn&, AppContext&) {
        Json payload = req.json();
        if (!payload.is_object()) {
            resp.set_error("json body required", 400);
            return;
        }
        std::string driver = payload.gets("driver", ctx.config.arm.backend);
        if (driver != ctx.config.arm.backend) {
            resp.set_error("driver mismatch", 400);
            return;
        }
        const Json* angles = payload.get("angles");
        if (!angles || !angles->is_object()) angles = &payload;
        std::string path = csrc::arm_angles_path();
        size_t slash = path.find_last_of('/');
        std::string def_file = (slash == std::string::npos ? "" : path.substr(0, slash + 1)) + "arm_angles_default.json";
        std::ofstream f(def_file);
        if (!f) {
            resp.set_error("cannot write default config", 500);
            return;
        }
        f << angles->dump(true, 0);
        Json j;
        j["status"] = "success";
        j["driver"] = driver;
        j["angles"] = *angles;
        resp.set_json(j);
    });

    // /api/arm/angles/preview
    router.add("POST", "/api/arm/angles/preview", [&ctx](const HttpRequest& req, HttpResponse& resp, ClientConn&, AppContext&) {
        Json payload = req.json();
        if (!payload.is_object()) {
            resp.set_error("json body is required", 400);
            return;
        }
        std::string driver = payload.gets("driver", ctx.config.arm.backend);
        if (driver != ctx.config.arm.backend) {
            resp.set_error("driver mismatch: expected " + ctx.config.arm.backend + ", got " + driver, 400);
            return;
        }
        std::string key = payload.gets("key");
        if (key.empty()) {
            resp.set_error("key is required", 400);
            return;
        }
        if (!payload.has("value")) {
            resp.set_error("value is required", 400);
            return;
        }
        const Json* angles = payload.get("angles");
        if (!angles || !angles->is_object()) {
            resp.set_error("angles is required", 400);
            return;
        }
        int value = (int)payload.geti("value", 0);

        Json normalized = csrc::save_arm_angles(driver, *angles);
        Json upd = update_arm_angles(ctx, driver, normalized);
        if (upd.has("error")) {
            resp.set_error(upd.gets("error"), 400);
            return;
        }
        Json prv = preview_arm_angle(ctx, driver, key, value);
        if (prv.has("error")) {
            resp.set_error(prv.gets("error"), 400);
            return;
        }
        Json j;
        j["status"] = "success";
        j["driver"] = driver;
        j["key"] = key;
        j["value"] = csrc::Json((int64_t)value);
        j["angles"] = normalized;
        resp.set_json(j);
    });
}
}  // namespace routes
}  // namespace capp
